File: src/main/scala/lib/Pos.scala
package lib

case class Pos(x: Int, y: Int) {
  def +(that: Pos): Pos = Pos(x + that.x, y + that.y)

  def *:(k: Int): Pos = Pos(k * x, k * y)

  def manhattanDistance(that: Pos): Int = (x - that.x).abs + (y - that.y).abs

  def <=(that: Pos): Boolean = x <= that.x && y <= that.y

  def min(that: Pos): Pos = Pos(x min that.x, y min that.y)

  def max(that: Pos): Pos = Pos(x max that.x, y max that.y)

  def getAdjacent: Seq[Pos] = List(Pos(x + 1, y), Pos(x - 1, y), Pos(x, y + 1), Pos(x, y - 1))
}

object Pos {
  val zero: Pos = Pos(0, 0)

  val axisOffsets: Seq[Pos] = Seq(Pos(0, 1), Pos(-1, 0), Pos(1, 0), Pos(0, -1))
  val diagonalOffsets: Seq[Pos] = Seq(Pos(-1, 1), Pos(1, 1), Pos(-1, -1), Pos(1, -1))
  val allOffsets: Seq[Pos] = axisOffsets ++ diagonalOffsets
}

File: csharp/fsharp/Pos.fs
module fsharp.Pos

type Pos = {
    X : int
    Y : int
} with

    // Operator overloads
    static member (+) (p1 : Pos, p2 : Pos) = {
        X = p1.X + p2.X
        Y = p1.Y + p2.Y
    }

    static member (-) (p1 : Pos, p2 : Pos) : Pos = {
        X = p1.X - p2.X
        Y = p1.Y - p2.Y
    }

    static member (*) (p1 : Pos, p2 : Pos) : Pos = {
        X = p1.X * p2.X
        Y = p1.Y * p2.Y
    }

    static member (<=) (p1 : Pos, p2 : Pos) : bool =
        p1.X <= p2.X && p1.Y <= p2.Y

    static member (>=) (p1 : Pos, p2 : Pos) : bool =
        p1.X >= p2.X && p1.Y >= p2.Y

module Pos =
    let create (x : int) (y : int) : Pos = {
        X = x
        Y = y
    }
    // Manhattan Distance
    let manhattanDistance (p1 : Pos) (p2 : Pos) : int =
        abs (p1.X - p2.X) + abs (p1.Y - p2.Y)

    // Axis Offsets
    let axisOffsets : List<Pos> = [
        {
            X = -1
            Y = 0
        }
        {
            X = 0
            Y = -1
        }
        {
            X = 0
            Y = 1
        }
        {
            X = 1
            Y = 0
        }
    ]

    // Diagonal Offsets
    let diagonalOffsets : List<Pos> = [
        {
            X = -1
            Y = -1
        }
        {
            X = -1
            Y = 1
        }
        {
            X = 1
            Y = -1
        }
        {
            X = 1
            Y = 1
        }
    ]

    let allOffsets = diagonalOffsets |> List.append axisOffsets

    // Get Axis Offsets
    let getAxisOffsets (pos : Pos) : List<Pos> =
        axisOffsets |> List.map (fun offset -> pos + offset)

    // Get Diagonal Offsets
    let getDiagonalOffsets (pos : Pos) : List<Pos> =
        diagonalOffsets |> List.map (fun offset -> pos + offset)

    // Get All Offsets
    let getAllOffsets (pos : Pos) : List<Pos> =
        List.append (getAxisOffsets pos) (getDiagonalOffsets pos)

    // Zero Position
    let zero : Pos = {
        X = 0
        Y = 0
    }
